### ISSUE TITLE
[feature] Add min_p sampling and frequency/presence penalties to generation

### ISSUE THREAD

**Issue by astorise** [huggingface/candle#3654]
### Motivation
`candle_transformers::generation::Sampling` offers ArgMax, All, TopK, TopP,
TopKThenTopP and GumbelSoftmax, but not **min_p** (a popular, robust alternative
to top_p) nor OpenAI-style **frequency_penalty / presence_penalty**. These are
expected by anyone exposing an OpenAI-compatible API on top of candle.

### Gap in candle
- No `min_p` variant in `Sampling`.
- No frequency/presence penalty helpers in the generation module (repeat-penalty,
  where present, lives outside this module and differs from OpenAI semantics).

### Proposed implementation
- Add `Sampling::MinP { p: f64, temperature: f64 }` (keep min-p tokens with prob
  ≥ `p * max_prob`), wired through `LogitsProcessor`.
- Add penalty helpers applied to logits before sampling, parameterized over the
  generated-token history, matching OpenAI's frequency/presence semantics.

### Use-case / example
```rust
// min_p sampling for an OpenAI-compatible server.
let mut lp = LogitsProcessor::from_sampling(seed, Sampling::MinP { p: 0.05, temperature: 0.8 });
let next = lp.sample(&logits)?;

// frequency / presence penalties over the running output.
let logits = apply_frequency_presence_penalty(&logits, &generated, freq: 0.5, presence: 0.3)?;
```

**Comment by RajaBabu15** [comment]
Opened PR #3731 implementing both pieces of this proposal:

- `Sampling::MinP { p, temperature }` wired through `LogitsProcessor` — keeps tokens with prob ≥ `p * max_prob`, samples from the renormalized survivors (matches the API sketched above)
- `apply_frequency_presence_penalty(logits, context, frequency_penalty, presence_penalty)` with OpenAI semantics: `logit[t] -= frequency_penalty * count(t) + presence_penalty`

Covered by unit tests (deterministic min-p filtering, renormalized-odds check over 10k draws, exact penalty arithmetic, no-op / out-of-vocab edges); clippy and fmt clean.

**Comment by astorise** [comment]
Thanks for jumping on this so quickly! The implementation looks well-scoped — Sampling::MinP threaded through LogitsProcessor and the OpenAI-compatible penalty formula are exactly what I had in mind.

A few things I'll check when reviewing [#3731](https://github.com/astorise/candle/issues/3731):

Min-p with temperature: does it apply temperature before or after the p-filter? (The paper recommends scaling first, then filtering on the rescaled distribution.)
Penalty reset: is there a way to clear the context token counts between generations, or is that left to the caller?
LogitsProcessor API: does it expose the new sampling mode via the existing builder, or does it require constructing a new processor?
Happy to test it on a couple of models once the CI is green. Great work!

**Comment by RajaBabu15** [comment]
Thanks for the quick look!

On temperature ordering — scaling happens first. The probs are computed as `softmax(logits/temperature)` and the min-p threshold is applied to that rescaled distribution, so it follows the paper.

The penalty helper is intentionally stateless. You hand it the generated-token slice each call, and it counts occurrences right there — nothing accumulates inside, so there's nothing to reset between generations—same contract as the existing `apply_repeat_penalty` in `utils`, just with OpenAI Math.

and no new processor — it's just `LogitsProcessor::from_sampling(seed, Sampling::MinP { p, temperature })`, same entry point as TopK/TopP. I left the `new(seed, temperature, top_p)` shorthand untouched since changing its signature would ripple through the examples.

Model runs once CI unblocks would be great — if the sampling behavior looks off on anything real, I'll fix it up.

**Comment by astorise** [comment]
All three points cleared up — temperature-first ordering is exactly right, the stateless penalty contract is clean (and consistent with apply_repeat_penalty), and keeping LogitsProcessor::from_sampling as the single entry point is the right call for backward compat.

I'll run it on a couple of models once CI is green and report back here.
